lib/AgentSDK.js

```diff
@@ -114,30 +148,113 @@ class SDK extends Events {
-                } else if (this.transport) {
-                    this.transport.configuration.token = token;
-                    this.transport.reconnect();
-                } else {
-                    // 3. init
-                    this._init(Object.assign({
-                        domain: domains.asyncMessagingEnt,
-                        token: token
-                    }, this.conf));
-                }
+            this._loginAndReconnect(domains, (err) => {
+                this._emitStatusEvents(err, 'Reconnect#Relogin#WS');
             });
+        });
+    }
 
+    /**
+     * Dispose all connections, event listeners, and timers connected to this agent
+     * You might want to call this function in the end of the agent life cycle
+     */
+    dispose() {
+        clearTimeout(this.errorTimeoutId);
+        clearTimeout(this.refreshSessionTimeoutId);
+        if (this.transport) {
+            this.transport.removeAllListeners('open');
+            this.transport.close();
+        }
+        this.removeAllListeners();
+        this.transport = null;
+        this.connected = false;
+    }
+
+    /**
+     * Start a periodic call to refresh the HTTP session, thus prolonging the lifetime of the bearer token
+     * The default time for the interval of call is 10 minutes.
+     */
+    startPeriodicRefreshSession() {
+        // Cancel all refresh session
+        clearTimeout(this.refreshSessionTimeoutId);
+
+        // Start periodic call
+        this.refreshSessionTimeoutId = setTimeout(() => {
+            this.refreshSession(() => {
+                // Re-call function after the last one passed
+                this.startPeriodicRefreshSession();
+            });
+        }, this.refreshSessionInterval);
+    }
+
+    /**
+     * Asynchronously refresh the HTTP session of the agent, thus prolonging the bearer token lifetime
+     * @param sessionCallback - function(err: Error) - the callback to check for errors
+     */
+    refreshSession(sessionCallback) {
+        async.waterfall([
+            // 1. Get CSDS domain
+            (callback) => this._handleRefreshSessionCSDS((err, domains) => {
+                callback(null, err, domains);
+            }),
+
+            // 2. Call Refresh Session
+            (err, domains, callback) => {
+                // Propagate error from CSDS right away
+                if (err) {
+                    return callback(null, err, domains);
+                }
+
+                this._handleRefreshSessionCall(domains, (err, domains) => {
+                    callback(null, err, domains);
+                });
+            },
+
+            // 3. Call re-login when 401 is encountered
+            (err, domains, callback) => {
+
+                // Re-connect on 401
+                if (err && err.code === 401) {
+                    this._handleRefreshSessionReconnect(domains, (err) => callback(err));
```

Comment:
Does this need a timeout to prevent spamming?

Reply:
Timeout in terms of what? like request timeout or setTimeout? it is every 10 minutes though

Reply:
Ah, ok i forgot that it was inside a setTimeout already above